File: solutions_python/hashmaps/202.happy-number.py
#
# @lc app=leetcode id=202 lang=python
#
# [202] Happy Number
#

# @lc code=start
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        seen = []
        while n > 0:
            res = 0
            while n > 0:
                res += (n % 10) ** 2
                n = n / 10
            if res in seen:
                return False
            if res == 1:
                return True
            seen.append(res)
            n = res
        return False
        
# @lc code=end

File: solutions_java/hashmaps/202.happy-number.java
package solutions_java.hashmaps;
/*
 * @lc app=leetcode id=202 lang=java
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> set = new HashSet<>();
        set.add(n);

        while (n != 1) {
            int copy = n;
            int res = 0;
            while (copy != 0) {
                res += (copy % 10) * (copy % 10);
                copy = copy / 10;
            }
            if (set.contains(res)) return false;
            set.add(res);
            n = res;
        }
        return true;
    }
}
// @lc code=end

